components/sync/driver/brave_sync_service_impl.cc

```diff
@@ -57,15 +59,30 @@ bool BraveSyncServiceImpl::IsSetupInProgress() const {
          !user_settings_->IsFirstSetupComplete();
 }
 
+void BraveSyncServiceImpl::StopAndClear() {
+  SyncServiceImpl::StopAndClear();
+  brave_sync_prefs_.Clear();
+}
+
 std::string BraveSyncServiceImpl::GetOrCreateSyncCode() {
-  std::string sync_code = brave_sync_prefs_.GetSeed();
+  bool failed_to_decrypt = false;
+  std::string sync_code = brave_sync_prefs_.GetSeed(&failed_to_decrypt);
+
+  if (failed_to_decrypt) {
+    // Do not try to re-create seed when OSCrypt fails, for example on macOS
+    // when the keyring is locked.
+    DCHECK(sync_code.empty());
+    return std::string();
```

Comment:
In this case, we're returning an error and an empty sync code.

Does the `GetOrCreateSyncCode()` caller check for this error and handle things differently?

Reply:
`GetOrCreateSyncCode` is used:
1. by Android code `BraveSyncWorker::GetSyncCodeWords`, expecting it is valid and non-empty:
```
  sync_code = sync_service->GetOrCreateSyncCode();

  CHECK(brave_sync::crypto::IsPassphraseValid(sync_code));

```
(https://github.com/brave/brave-core/blob/master/browser/android/brave_sync_worker.cc#L97)

2. By desktop UI page
```BraveSyncHandler::HandleGetSyncCode
<=
SyncSetupGetSyncCode
<=
getSyncCode()
```
this is invoked in cases of show codewords, create sync chain. No any verification of `GetOrCreateSyncCode`'s result happens.

3. iOS code 
If I understand correctly, when it shows up `connect with codewords` dialog - `GetOrCreateSyncCode`'s result is ignored, and when it shows up `connect through QR` then there is a `PassphraseToBytes32` verification.
(https://github.com/brave/brave-core/blob/master/ios/browser/api/sync/brave_sync_api.mm#L114 and https://github.com/brave/brave-core/blob/master/ios/browser/api/sync/brave_sync_api.mm#L138)

Reply:
iOS does the isValid check in Swift instead of checking in C++. We have the check in the QRCodeModel line 147: https://github.com/brave/brave-ios/blob/development/Client/Frontend/Sync/BraveCore/BraveSyncQRCodeModel.swift#L147

And L196 when scanning via Camera: https://github.com/brave/brave-ios/blob/development/Client/Frontend/Sync/SyncPairCameraViewController.swift#L196

And L204 when entering it manually: https://github.com/brave/brave-ios/blob/development/Client/Frontend/Sync/SyncPairWordsViewController.swift#L204

))

Reply:
> 3\. By desktop UI page
> 
> ```
> <=
> SyncSetupGetSyncCode
> <=
> getSyncCode()
> ```
> 
> this is invoked in cases of show codewords, create sync chain. No any verification of `GetOrCreateSyncCode`'s result happens.

So if I understand you correctly this means that in case of failure here we would show a QR code for a Sync code which consists of only the empty string?

Reply:
@fmarier 
Yes. Here is what is displayed when the chain was created but then something happened and `GetOrCreateSyncCode` gives error:
![Screenshot from 2022-01-28 15-53-48](https://user-images.githubusercontent.com/24739341/151559233-2879b97d-3403-4e7a-8299-da0e47f32397.png)
![Screenshot from 2022-01-28 15-53-36](https://user-images.githubusercontent.com/24739341/151559237-65e2e142-c7af-44a4-a09c-dfc24767ab72.png)
![Screenshot from 2022-01-28 15-53-23](https://user-images.githubusercontent.com/24739341/151559239-24828f73-4344-474c-9138-2e26c718b957.png)